src/main/java/com/mraof/minestuck/block/redstone/TogglerBlock.java

```diff
@@ -0,0 +1,148 @@
+import net.minecraft.world.World;
+import net.minecraftforge.common.util.Constants;
+
+/**
+ * Toggles the state of the block it is facing upon receiving a new redstone signal if that block has the MACHINE_TOGGLE property from MSProperties,
+ * or returns redstone power in the form of the POWERED/POWER properties to false/0 if DISCHARGE is set to true
+ */
+public class TogglerBlock extends MSDirectionalBlock
+{
+	public static final BooleanProperty POWERED = BlockStateProperties.POWERED;
+	public static final BooleanProperty DISCHARGE = MSProperties.MACHINE_TOGGLE;
+	
+	public TogglerBlock(Properties properties)
+	{
+		super(properties);
+		this.registerDefaultState(stateDefinition.any().setValue(FACING, Direction.UP).setValue(POWERED, false).setValue(DISCHARGE, false));
+	}
+	
+	@Override
+	@SuppressWarnings("deprecation")
+	public ActionResultType use(BlockState state, World worldIn, BlockPos pos, PlayerEntity player, Hand hand, BlockRayTraceResult hit)
+	{
+		worldIn.setBlock(pos, state.cycle(DISCHARGE), Constants.BlockFlags.DEFAULT);
+		if(state.getValue(DISCHARGE))
+			worldIn.playSound(null, pos, SoundEvents.PISTON_EXTEND, SoundCategory.BLOCKS, 0.5F, 1.2F);
+		else
+			worldIn.playSound(null, pos, SoundEvents.PISTON_CONTRACT, SoundCategory.BLOCKS, 0.5F, 1.2F);
+		return ActionResultType.SUCCESS;
+	}
+	
+	@Override
+	public void neighborChanged(BlockState state, World worldIn, BlockPos pos, Block blockIn, BlockPos fromPos, boolean isMoving)
+	{
+		super.neighborChanged(state, worldIn, pos, blockIn, fromPos, isMoving);
+		if(fromPos != pos.relative(state.getValue(FACING)))
+			updatePower(worldIn, pos);
+	}
+	
+	@Override
+	public void onPlace(BlockState state, World worldIn, BlockPos pos, BlockState oldState, boolean isMoving)
+	{
+		super.onPlace(state, worldIn, pos, oldState, isMoving);
+		updatePower(worldIn, pos);
+	}
+	
+	public void updatePower(World worldIn, BlockPos pos)
+	{
+		if(!worldIn.isClientSide)
+		{
+			BlockState state = worldIn.getBlockState(pos);
+			boolean hasPower = false;
+			Direction stateFacing = state.getValue(FACING);
+			
+			for(Direction direction : Direction.values()) //checks for a signal in any direction except the one it is facing
+			{
+				if(direction != stateFacing && worldIn.getSignal(pos.relative(direction), direction) > 0)
+				{
+					hasPower = true;
+					break;
+				}
+			}
+			
+			boolean isPoweredBeforeUpdate = state.getValue(POWERED);
+			
+			if(state.getValue(POWERED) != hasPower)
+				worldIn.setBlockAndUpdate(pos, state.setValue(POWERED, hasPower));
+			else worldIn.sendBlockUpdated(pos, state, state, 2);
+			
+			if(!isPoweredBeforeUpdate && hasPower)
+			{
+				BlockPos facingPos = pos.relative(state.getValue(FACING));
+				BlockState facingState = worldIn.getBlockState(facingPos);
+				
+				if(state.getValue(DISCHARGE))
+				{
+					boolean makeSound = false;
+					
+					//removing redstone power from powered blocks
+					for(Property<?> property : facingState.getProperties())
```

Comment:
Rewriting this section to use `facingState.hasProperty()` should make this section much simpler.